pkg/apis/resource/validation/validation.go

```diff
@@ -682,14 +685,28 @@ func stringKey(item string) (string, string) {
 
 // validateMap validates keys, items and the maximum length of a map.
 // A negative maxSize disables the length check.
-func validateMap[K ~string, T any](m map[K]T, maxSize int, validateKey func(K, *field.Path) field.ErrorList, validateItem func(T, *field.Path) field.ErrorList, fldPath *field.Path) field.ErrorList {
+//
+// Keys larger than maxKeyLen get truncated in the middle.
+func validateMap[K ~string, T any](m map[K]T, maxSize, maxKeyLen int, validateKey func(K, *field.Path) field.ErrorList, validateItem func(T, *field.Path) field.ErrorList, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
 	if maxSize >= 0 && len(m) > maxSize {
 		allErrs = append(allErrs, field.TooLongMaxLength(fldPath, len(m), maxSize))
 	}
 	for key, item := range m {
-		allErrs = append(allErrs, validateKey(key, fldPath)...)
-		allErrs = append(allErrs, validateItem(item, fldPath.Key(string(key)))...)
+		keyPath := fldPath.Key(truncateIfTooLong(string(key), maxKeyLen))
+		allErrs = append(allErrs, validateKey(key, keyPath)...)
+		allErrs = append(allErrs, validateItem(item, keyPath)...)
 	}
 	return allErrs
 }
+
+func truncateIfTooLong(str string, maxLen int) string {
+	if len(str) < maxLen {
+		return str
+	}
+	if maxLen <= 3 {
+		return strings.Repeat(".", maxLen)
```

Comment:
> if the len is short, why not just return the string?

Isn't that what `if len(str) < maxLen { return str }` does?

> And for short maxlen, we should allow to print more

It's already in the range of 100 characters, that won't make a different the way it is used now. It would in other circumstances.

Reply:
> Isn't that what if len(str) < maxLen { return str } does?

That's the "valid" path.  I am saying that very short stings should be printed fully, even if they are longer than is allowed.  E.g. if the limit is 16, and someone passes 18 characters, there's no point in getting fancy.

> It's already in the range of 100 characters, that won't make a different the way it is used now. It would in other circumstances.

I guarantee this pattern gets copied :)

Reply:
I don't think that `truncateIfTooLong` or `validateMap` should second-guess their caller. I added some comment to `validateMap` about choosing a higher truncation limit if valid keys are very short.

Reply:
This technique was supposed to be for humans, balancing the risk of printing stupidly long errors with the probability of including useful parts of the string in question.

I get that you don't want to second-guess the caller, but looking at the test, shows a few things, exclusively at the "small" end.

"hello" -> "h...o" (max 5) is utterly pointless (and I think wrong, actually).

"hello" -> "h..." (max 4) is similarly pointless.

"hello" -> "..." (max 3) is somehow worse than pointless - negative points!

If the user said "abc" but maxLen was 2, there's no world where ".." is a better error message.

My point is that, if the length is "short" (which is arbitrary) we don't really need to abbreviate at all.  The harm done by printing "hello" when the limit is 3 is non-existent.

It seems unlikely that we have any places that have limits that low, but I don't like leaving the logic this way.  I'd much rather document this as "this will elide the middle of large values, but small values will be retained, even if they exceed the maxLen."  If you want to define small as 8 or 10, that's fine with me.

Reply:
> looking at the test, 

Those are unit tests and entirely artificial. I'll change the code as you suggested, although it makes no difference.

Reply:
Done and squashed into "DRA API: enhance validation testing".